src/modules/firewall/src/lib/Firewall.h

```diff
@@ -39,9 +40,128 @@ class FirewallLog
     static OSCONFIG_LOG_HANDLE m_logHandle;
 };
 
+class GenericRule
+{
+public:
+    enum class State
+    {
+        Present,
+        Absent
+    };
+
+    enum class Action
+    {
+        Accept,
+        Drop,
+        Reject
+    };
+
+    enum class Direction
+    {
+        In,
+        Out
+    };
+
+    enum class Protocol
+    {
+        Any,
+        Tcp,
+        Udp,
+        Icmp,
+    };
+
+    GenericRule() = default;
+    virtual ~GenericRule() = default;
+
+    virtual GenericRule& Parse(const rapidjson::Value& rule);
+
+    virtual std::vector<std::string> GetParseError() const
+    {
+        return m_parseError;
+    }
+
+    virtual bool HasParseError() const
+    {
+        return !m_parseError.empty();
+    }
+
+    virtual State GetDesiredState() const
+    {
+        return m_desiredState;
+    }
+
+    virtual std::string Specification() const = 0;
+
+    virtual int ActionFromString(const std::string& str);
+    virtual int DirectionFromString(const std::string& str);
+    virtual int StateFromString(const std::string& str);
+    virtual int ProtocolFromString(const std::string& str);
+
+    virtual std::string ActionToString() const = 0;
+    virtual std::string DirectionToString() const = 0;
+    virtual std::string ProtocolToString() const = 0;
+
+protected:
+    Action m_action;
+    Direction m_direction;
+    Protocol m_protocol;
+    std::string m_sourceAddress;
+    std::string m_destinationAddress;
+    std::string m_sourcePort;
+    std::string m_destinationPort;
+
+    std::vector<std::string> m_parseError;
+
+private:
+    State m_desiredState;
+};
+
+class IpTablesRule : public GenericRule
+{
+public:
+    IpTablesRule() = default;
+    virtual ~IpTablesRule() = default;
+    virtual std::string Specification() const override;
+
+    virtual std::string ActionToString() const override;
+    virtual std::string DirectionToString() const override;
+    virtual std::string ProtocolToString() const override;
+
+    virtual std::string ActionToTarget() const;
+    virtual std::string DirectionToChain() const;
+
+    virtual int ActionFromTarget(const std::string& str);
+    virtual int DirectionFromChain(const std::string& str);
+};
+
+class IpTablesPolicy : public IpTablesRule
+{
+public:
+    typedef GenericRule::Action Action;
+    typedef GenericRule::Direction Direction;
+
+    IpTablesPolicy() = default;
```

Comment:
There shouldn't be a need to explicitly define these default functions, is there?

Reply:
Good point! I will remove all these default constructors